### PR TITLE
223 Refresh verification QR code with changing OTP

### LINKED ISSUE #223: Refresh verification QR code with changing OTP
Refers to: #176. Followup of #16.

- Change qr code at the correct time
- Visualize the remaining time for the current qr code

### PR DESCRIPTION
**Pull request by PaulaWi**
closes #223

### CODE REVIEW

**Comment by RoAnBu** [comment]
This error gets spammed into the console when viewing the qr code on ios
```
======== Exception caught by animation library =====================================================
The following assertion was thrown while notifying listeners for AnimationController:
'package:flutter/src/widgets/framework.dart': Failed assertion: line 4263 pos 12: '_debugLifecycleState != _ElementLifecycle.defunct': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=BUG.md

When the exception was thrown, this was the stack: 
#2      Element.markNeedsBuild (package:flutter/src/widgets/framework.dart:4263:12)
#3      State.setState (package:flutter/src/widgets/framework.dart:1264:14)
#4      _AnimatedProgressbarState.initState.<anonymous closure> (package:ehrenamtskarte/identification/animated_progressbar.dart:27:9)
#5      AnimationLocalListenersMixin.notifyListeners (package:flutter/src/animation/listener_helpers.dart:136:19)
#6      AnimationController._tick (package:flutter/src/animation/animation_controller.dart:797:5)
...
The AnimationController notifying listeners was: AnimationController#926bd(▶ 0.688)
====================================================================================================

======== Exception caught by animation library =====================================================
'package:flutter/src/widgets/framework.dart': Failed assertion: line 4263 pos 12: '_debugLifecycleState != _ElementLifecycle.defunct': is not true.
====================================================================================================
```

**Review by elKei24 — APPROVED** [review]
Did you try it in landscape mode?

I did not execute it to see how it looks. In case you want me to do so before merging, tell me.

File: frontend/lib/identification/animated_progressbar.dart [hunk 1]
```diff
@@ -21,7 +21,9 @@ class _AnimatedProgressbarState extends State<AnimatedProgressbar>
   void initState() {
     super.initState();
     controller = AnimationController(
-        duration: Duration(milliseconds: widget.duration), vsync: this);
+        duration: Duration(
+            milliseconds: widget.duration > 0 ? widget.duration : 1 << 63),
```
Comment (elKei24) [inline frontend/lib/identification/animated_progressbar.dart:25]: What does this bitshift? Maybe we could have a named constant for that. Or can we just leave away the animation if the duration is leq 0?

**Review by elKei24 — APPROVED** [review]

File: frontend/lib/identification/otp_generator.dart [hunk 2]
```diff
@@ -9,11 +12,28 @@ class OTPGenerator {
...
-        _base32TotpSecret, DateTime.now().millisecondsSinceEpoch,
-        algorithm: _algorithm,
-        length: _otpLength,
-        interval: _otpIntervalSeconds));
+  OTPCode generateOTP([VoidCallback onTimeout]) {
+    final time = DateTime.now().millisecondsSinceEpoch;
+    final validUntilMilliSeconds =
+        ((((time ~/ 1000).round()) ~/ _otpIntervalSeconds).floor() + 1) *
+            _otpIntervalSeconds *
+            1000;
```
Comment (Schedulaar) [inline frontend/lib/identification/otp_generator.dart:20]: ~/ is the [truncate operator](https://api.flutter.dev/flutter/dart-core/double/operator_truncate_divide.html), so I think, the following should suffice
```suggestion
    final intervalMilliSeconds = _otpIntervalSeconds * 1000;
    final validUntilMilliSeconds = (time ~/ intervalMilliseconds + 1) * intervalMilliSeconds;
```

File: frontend/lib/identification/animated_progressbar.dart [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  void initState() {
+    super.initState();
+    controller = AnimationController(
+        duration: Duration(
+            milliseconds: widget.duration > 0 ? widget.duration : 1 << 63),
+        vsync: this);
+    animation = Tween(begin: 1.0, end: 0.0).animate(controller)
+      ..addListener(() {
+        setState(() {});
+      })
```
Comment (Schedulaar) [inline frontend/lib/identification/animated_progressbar.dart:30]: What does that do?

File: frontend/lib/identification/verification_qr_code_view.dart [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import 'animated_progressbar.dart';
+import 'card_details.dart';
+import 'card_details_model.dart';
+import 'otp_generator.dart';
+
+class VerificationQrCodeView extends StatefulWidget {
+  final CardDetails cardDetails;
+  final OTPGenerator _otpGenerator;
+  static const otpResetIntervalSeconds = 30;
+  static const otpLength = 10;
```
Comment (Schedulaar) [inline frontend/lib/identification/verification_qr_code_view.dart:18]: We go for 6 digits!
```suggestion
  static const otpLength = 6;
```

**Review by Schedulaar — APPROVED** [review]

File: frontend/lib/identification/animated_progressbar.dart (hunk 3 above)
Comment (PaulaWi) [reply frontend/lib/identification/animated_progressbar.dart:30]: It updates the widget when the progress value changes

File: frontend/lib/identification/animated_progressbar.dart (hunk 1 above)
Comment (PaulaWi) [reply frontend/lib/identification/animated_progressbar.dart:25]: I will leave away the animation. There is no max integer constant in dart, thats why I used the bit shift.

**Pull request merged**